Initialize codejail_service application

timmc-edx opened the pull request:
Two commits, best reviewed individually:

- Straight-up output from a cookiecutter run
- A fairly large commit that mostly deletes unwanted things and updates dependencies (but see its full description for many details)

This is part of https://github.com/edx/edx-arch-experiments/issues/894

robrap commented on codejail_service/settings/base.py:
```diff
@@ -26,218 +24,40 @@ def root(*path_fragments):
-    'edx_django_utils.monitoring.CookieMonitoringMiddleware',  # cookie names (compliance) and sizes
-    'edx_django_utils.monitoring.CachedCustomMonitoringMiddleware',  # support accumulate & increment
-    'edx_django_utils.monitoring.MonitoringMemoryMiddleware',  # memory usage
+    # TODO: re-enable once we have waffle available
+    # 'edx_django_utils.monitoring.DeploymentMonitoringMiddleware',  # python and django version
+    # 'edx_django_utils.monitoring.CookieMonitoringMiddleware',  # cookie names (compliance) and sizes
+    # 'edx_django_utils.monitoring.CachedCustomMonitoringMiddleware',  # support accumulate & increment
+    # 'edx_django_utils.monitoring.MonitoringMemoryMiddleware',  # memory usage
```
Changes I made below:
1. I updated the TODO. I don't see how the original made sense, but I might be missing something.
2. I removed CookieMonitoringMiddleware.
3. I added proposed priorities for re-enabling into the comments. 
 
```
    # TODO: re-enable once waffle. which depends on a db, is optional. <INSERT TICKET LINK>

    # 'edx_django_utils.monitoring.DeploymentMonitoringMiddleware',  # python and django version (Re-enable Priority: P3) 
    # 'edx_django_utils.monitoring.CachedCustomMonitoringMiddleware',  # support accumulate & increment (Re-enable Priority: P5) 
    # 'edx_django_utils.monitoring.MonitoringMemoryMiddleware',  # memory usage (Re-enable Priority: P5)
```

robrap replied:
@timmc-edx: Please put 0 hours against resolving anything waffle related until we get a chance to speak on this topic. Thank you.

robrap commented on codejail_service/settings/base.py:
```diff
@@ -26,218 +24,40 @@ def root(*path_fragments):
-    # Resets RequestCache utility for added safety.
-    'edx_django_utils.cache.middleware.RequestCacheMiddleware',
-
     # Monitoring middleware should be immediately after RequestCacheMiddleware
```
This comment is no longer relevant.

timmc-edx replied:
Thanks, removed comment.

robrap commented on codejail_service/settings/base.py:
```diff
@@ -26,218 +24,40 @@ def root(*path_fragments):
-BACKEND_SERVICE_EDX_OAUTH2_KEY = 'replace-me'
-BACKEND_SERVICE_EDX_OAUTH2_SECRET = 'replace-me'
-
-JWT_AUTH = {
```
Is the eduNEXT version entirely without auth? Is it simply trusting the behavior of its caller?

timmc-edx replied:
Correct -- theirs is an unauthenticated service as well. I originally wanted to protect ours behind JWT auth but decided it wasn't worth it; anyone can already make calls to the service as they wish by submitting answers to certain course problems. And we can just decline to expose a direct API to the internet anyhow.

robrap commented on codejail_service/settings/test.py:
```diff
@@ -2,15 +2,10 @@
 from codejail_service.settings.base import *  # pylint: disable=wildcard-import
 
-# IN-MEMORY TEST DATABASE
+# Django *really* wants a database. So we give it a blank in-memory one.
```
Dang. Really? I wonder if we could rig it so it blows up if any unit test actually tries to use it?

timmc-edx replied:
I'll go ahead and move it to base settings for consistency.

robrap commented on codejail_service/urls.py:
```diff
@@ -19,30 +19,12 @@
-    re_path(r'^admin/', admin.site.urls),
+urlpatterns = [
     re_path(r'^api/', include(api_urls)),
-    re_path(r'^api-docs/', get_swagger_view(title='codejail-service API')),
```
Could this be useful in a development environment?

timmc-edx replied:
Hmm... I suppose. I could bring some of these things back in for consistency with other IDAs.

timmc-edx replied:
Looks like this maybe requires a DB.

robrap commented on requirements/base.in:
```diff
@@ -2,17 +2,9 @@
 Django         # Web application framework
-django-cors-headers
-django-extensions
-django-rest-swagger
```
Again - a question mark.

robrap commented on requirements/base.in:
```diff
@@ -2,17 +2,9 @@
-django-extensions
-django-rest-swagger
-django-waffle
-djangorestframework
```
I don't know if DRF will still make sense for consistency or not.

robrap commented on requirements/base.in:
```diff
@@ -2,17 +2,9 @@
+gunicorn       # HTTP server framework
+
 edx-django-utils
-edx-django-release-util
```
Do you know what this is for?

timmc-edx replied:
No, and the README was pretty vague:

> Release pipeline utilities for edX independently-deployable applications (IDAs) based on Django.

The `__init__.py` said this:

> a collection of Django management commands used for analyzing and manipulating migrations.

...and then it just has a bunch of management commands relating to databases.

timmc-edx replied:
(So because it was all DB-related, I just removed it.)

robrap reviewed (commented):
Thanks for the work on this one.

timmc-edx commented on README.rst:
```diff
@@ -0,0 +1,145 @@
+codejail_service
```
Various notes from Slack:

>  ... being exposed to the outside world (cors, csrf).

TODO: Reminder to document this and to communicate this to SRE.

> Add a blank DB for tests.

[question] May not be worth it, but is this something we can clean-up later? Seems a little risky that something like use of a waffle flag could make it through a unit test. Maybe we create a separate time-boxed ticket?

> Remove unneeded requirements files. We're going to deploy the base layer instead of the production layer, and requirements.txt is just not needed at all as far as I can tell.

Can you help me understand this? Is this an issue unique to this new service, or does cookie-cutter need to change? Other?

> Remove Dockerfile and devstack.py as these will be provided elsewhere.

I wonder where we (the community) are with this? I thought they wanted some basics. Should we find out, or should we err on removing what we won't use and having the maintainer add it when and if it is needed? It'll still be in git commit history.

> Re-enable various edx-django-utils middleware as soon as I can determine how to get them working without a DB. (Some of them use waffle flags.)

Can we task this in an entirely separate ticket so we can prioritize against other work? We might want an issue in edx-django-utils, even if we have an experiments ticket. (edited)

timmc-edx replied:
- Documented and communicated
- DB is now the same for all settings
- For requirements.txt, I'd like to change the cookiecutter. For production.in... not sure. I'm not clear on why we'd have these additional dependencies that we wouldn't use in base, where it would then be inherited by dev/test/etc. So I just moved gunicorn into the base and deleted production. (I'm still a little unclear on whether we want gevent. Some IDAs have it, some don't. I think it's maybe useful mostly for IO-bound loads?)
- (Docker/devstack covered elsewhere in comments)
- Tasked out.

timmc-edx commented on README.rst:
```diff
@@ -0,0 +1,145 @@
+
+.. note::
+
+  This README was auto-generated. Maintainer: please review its contents and
+  update all relevant sections. Instructions to you are marked with
+  "PLACEHOLDER" or "TODO". Update or remove those sections, and remove this
+  note when you are done.
```
Still need to get the README filled out properly.

timmc-edx commented on Makefile:
```diff
@@ -142,60 +129,11 @@ pull_translations:
 
 validate_translations: fake_translations detect_changed_source_translations ## install fake translations and check if translation files are up-to-date
 
-docker_build:
```
Here's the devstack DEPR: https://github.com/openedx/public-engineering/issues/247

It describes removing all of the devstack.py from repos, and contemplates removing all of the docker elements as well, since it's unused in Tutor. I think we want to get out ahead of that.

robrap replied:
Start reading the conversation starting here: https://github.com/openedx/public-engineering/issues/247#issuecomment-2344711918

robrap replied:
@timmc-edx: I don't think this conversation is resolved yet.
1. Please see my last comment.
2. I don't think this requires any changes to your PR, but I do think it requires a follow-up communication/discussion somewhere about this open question. Let me know if you will initiate that, or if you need me to. Thank you.

timmc-edx replied:
Oh, I misunderstood the meaning of your previous comment. I see now that there might be a role for a base development file that Tutor and Devstack can both build on. But it looks like Diana is already involved in the broader discussion?

robrap replied:
Sure. I don't know how this (or any other) new service should be treated at this time, and who might be keeping track (if needed) of the services that will need to be updated once they come to resolution.

timmc-edx commented on codejail_service/docker_gunicorn_configuration.py:
```diff
@@ -5,7 +5,6 @@
 
 preload_app = True
 timeout = 300
-bind = "0.0.0.0:18030"
```
Good idea -- apparently we can configure this via codecov.yml.

timmc-edx commented on Makefile:
```diff
@@ -79,14 +74,7 @@ lint: ## run Python code linting
-validate: test quality pii_check check_keywords ## run tests, quality, and PII annotation checks
+validate: test quality ## run tests, quality
 
 migrate: ## apply database migrations
```
Ticketed in https://github.com/openedx/codejail-service/issues/4

robrap commented on codejail_service/settings/base.py:
```diff
@@ -45,8 +41,13 @@ def root(*path_fragments):
 
-# No database for codejail
-DATABASES = {}
+# Django *really* wants a database. So we give it a blank in-memory one.
```
Do we want to add any sort of warning in the README that the service should be treated as-if it has no database?

timmc-edx replied:
It's mentioned in the README -- do you think it should be more prominent?

robrap replied:
Yes. And I don't just mean that the service does not have a database, but a warning about the required db for Django. For example, something like:
```
Warning: This service is configured with an in-memory database simply to make Django happy. The service itself should be considered to not to have a database, and should not rely on any database-dependent features, like waffle-based toggles, as one example.

timmc-edx replied:
Ah, I see. Made some changes, take a look.

robrap reviewed (approved):
Minor comments, including communication questions.

timmc-edx commented:
Rewrote commits to squash fixups but preserve the cookiecutter/changes distinction, and edit commits. Cleaned up license file move/move-back in commit history.

Only change to diff is that the license file no longer includes the "APPENDIX: How to apply the Apache License to your work." section that we usually omit.
